### ISSUE TITLE
Server side API documentation of tokenSwapURL and tokenRefreshURL in token swap service?

### ISSUE THREAD

**Issue by matsev** [spotify/ios-sdk#695]
What should the requests and responses to `tokenSwapURL` and `tokenRefreshURL` look like? Please provide information about request headers (e.g. `Content-Type` and `Accept`), request parameters (if any), request body (content type, key names in the payload, JSON and/or URI encoding, etc), response body (content type, key names, JSON and/or URI encoding, etc).

In the [documentation](https://developer.spotify.com/technologies/spotify-ios-sdk/tutorial/#setting-up-your-token-exchange-service) it is stated that:

> In the method we ask SPTAuth to check if the URL given is a Spotify authentication callback and, if so, we call your token swap service (see the Readme in the download package for more information on this) to get an access token.

However, the Readme of [ios-sdk-deploy-beta-17.zip](https://github.com/spotify/ios-sdk/releases/download/beta-17/ios-sdk-deploy-beta-17.zip) does not contain this information.

So far I have tried using request parameters:
```
curl - X POST https://my-server.com/token/swap?client_id=abc&code=xyz&redirect_uri=http%3A%2F%2Fmy-app.com%2Fcallback
```

and JSON payload:
```
curl -H "Content-Type: application/json" -X POST -d '{"client_id":"abc","code":"xyz", "redirect_uri":"http://my-app.com/callback"}' https://my-server.com/token/swap
```

**Comment by jakereps** [comment]
The [ios-sdk/spotify_towen_swap.rb](/spotify/ios-sdk/blob/master/Demo%20Projects/spotify_token_swap.rb#L58-L61) says in the comments it only takes a single parameter, and combines it with all the information that should be in the file already.

```ruby
post '/swap' do

    # This call takes a single POST parameter, "code", which
    # it combines with your client ID, secret and callback
    # URL to get an OAuth token from the Spotify Auth Service,
    # which it will pass back to the caller in a JSON payload.

    auth_code = params[:code]
```

and refresh takes just the refresh token:

```ruby
post '/refresh' do

    # Request a new access token using the POST:ed refresh token

    ...

    encrypted_token = params[:refresh_token]
```

Why try and hit it yourself when the SDK will do it for you?

Refresh is through:

[`– renewSession:callback:`](https://developer.spotify.com/ios-sdk-docs/Documents/Classes/SPTAuth.html#//api/name/renewSession:callback:)


Then about half way [down in the tutorial it says:](https://developer.spotify.com/technologies/spotify-ios-sdk/tutorial/)

> In the method we ask SPTAuth to check if the URL given is a Spotify authentication callback and, if so, **we call your token swap service** (see the Readme in the download package for more information on this) to get an access token. Once we have the access token, we call the -playUsingSession: method to play a track.

talking about [`– handleAuthCallbackWithTriggeredAuthURL:callback:`](https://developer.spotify.com/ios-sdk-docs/Documents/Classes/SPTAuth.html#//api/name/handleAuthCallbackWithTriggeredAuthURL:callback:)

**Comment by matsev** [comment]
@jakereps Thanks for your answer, I was not aware of the [ios-sdk/spotify_token_swap.rb](https://github.com/spotify/ios-sdk/blob/master/Demo%20Projects/spotify_token_swap.rb). We do not currently use Ruby and since we do not want to add more programming languages and frameworks to our backend stack than what we currently have we need to port the implementation.

Some questions have been answered by `spotify_,token_swap.rb`. Based on the source code I interpret that the responses should basically be a pass through of the response from the Spotify account service, and that they should be in JSON. 

However, there is still some information missing:
- In the [comment of `/swap`](https://github.com/spotify/ios-sdk/blob/master/Demo%20Projects/spotify_token_swap.rb#L58-L61) it is stated that ot has a single `code` parameter, but it does not say which content type is used. Is it JSON, XML, form data, etc? 
- Likewise, I assume that the same content type is used for `/refresh`, but it it does not say which.

Other thoughts:
- By hardcoding the `CLIENT_ID` we must deploy one server for each client id that is used. If the iOS SDK sends the client id as part of the request, we can have _one_ server instance that maps many client ids to their corresponding client secrets before calling the Spotify account service.

- By hardcoding the `CLIENT_CALLBACK_URL` in the server needs to share one more value with the client. It is not sensitive information and if the client passes its value then the server it can just forward it when making the request to `https://accounts.spotify.com/api/token` which would make the solution less fragile.
- What is the purpose of [encryption of the refresh token](https://github.com/spotify/ios-sdk/blob/master/Demo%20Projects/spotify_token_swap.rb#L80)? Since we do not give away the client secret the hacker cannot authenticate to the Spotify account service on our behalf. If we simply pass through the returned refresh token from the Spotify account service (without additional encryption) to the app, all an evil hacker could do with it is to call our web service again to refresh it.

**Comment by jakereps** [comment]
> In the comment of /swap it is stated that ot has a single code parameter, but it does not say which content type is used. Is it JSON, XML, form data, etc?

I'm pretty sure it is JSON, but you could always put in some fake uris, and pull up [Instruments to catch the requests sent by the SDK.](https://developer.apple.com/library/prerelease/content/documentation/DeveloperTools/Conceptual/InstrumentsUserGuide/MeasuringIO.html)

> By hardcoding the CLIENT_ID we must deploy one server for each client id that is used.

The `CLIENT_ID` is a per App identifier given to you by registering an App on Spotify's Developer page, you only need one unless you're building multiple Spotify Apps. 

I'm unsure of what you mean by the second bullet point.

> What is the purpose of encryption of the refresh token?

I was actually wondering the same thing when looking into it yesterday. Unless Spotify is just trying to make it harder to figure out their token generation system by hashing the tokens in some way, but I'm not really sure.


If you want some other swap files to look at there are PHP, python, and more in various repos around GitHub. It seems to work to do a search for 'Spotify token swap (python | php | ...)' on Google.

**Comment by matsev** [comment]
@jakereps We also assumed that the payload was JSON, but more tests revealed that the content type is `application/x-www-form-urlencoded` for the `/swap` endpoint (the client developers has left for vacation, so we have not checked the `/refresh` endpoint yet).

> The CLIENT_ID is a per App identifier given to you by registering an App on Spotify's Developer page, you only need one unless you're building multiple Spotify Apps.

Quite right. By providing the `CLIENT_ID` as part of the request, one server can serve multiple Spotify Apps.

> I'm unsure of what you mean by the second bullet point.

Sorry, I intended to write `REDIRECT_URI` (I have corrected my comment). Copied from [4. Your application requests refresh and access tokens](https://developer.spotify.com/web-api/authorization-guide/):

> redirect_uri _Required._ This parameter is used for validation only (there is no actual redirection). The value of this parameter must _exactly_ match the value of redirect_uri supplied when requesting the authorization code.

My point is that if the client pass the `REDIRECT_URI` as an argument instead of it being hardcoded in both the client and the server, one source of error can be eliminated.
